sky/packages/sky/lib/widgets/framework.dart

```diff
@@ -975,24 +976,29 @@ abstract class RenderObjectWrapper extends Widget {
     // we give them their own slots for them to fit into us.
   }
 
+  /// Override this function if your RenderObjectWrapper uses a [ParentDataNode]
+  /// to program parent data into children.
+  void updateParentData(RenderObject child, ParentDataNode node) { }
+
   void syncRenderObject(RenderObjectWrapper old) {
     assert(old == null || old.renderObject == renderObject);
-    ParentData parentData = null;
-    Widget ancestor = parent;
-    while (ancestor != null && ancestor is! RenderObjectWrapper) {
-      if (ancestor is ParentDataNode && ancestor.parentData != null) {
-        if (parentData != null)
-          parentData.merge(ancestor.parentData); // this will throw if the types aren't the same
-        else
-          parentData = ancestor.parentData;
+    ParentDataNode parentDataNode = null;
+    for (Widget current = parent; current != null; current = current.parent) {
+      assert(() {
+        if (current is ParentDataNode) {
+          Widget parentWrapper = current;
+          while (parentWrapper is! RenderObjectWrapper)
+            parentWrapper = parentWrapper.parent;
+          current.debugValidateParent(parentWrapper);
+        }
+        return true;
+      });
+      if (current is ParentDataNode) {
+        parentDataNode = current;
+      } else if (current is RenderObjectWrapper) {
+        current.updateParentData(renderObject, parentDataNode);
+        break;
```

Comment:
So this is interesting. It means that a Component can have a build() function that returns a ParentDataNode, for instance. It also means that a ParentDataNode will get ignored if there's another ancestor ParentDataNode.

I think we should say that the parent of a ParentDataNode must be the RenderObjectWrapper.